### PR TITLE
Adding Struct Support

### LINKED ISSUE #65: Appears that Struct is not a supported data type [BUG]
**What is the bug?**
Attempted to extract Struct (by calling `resultSet.getObject(<fieldName>)`) on returned ResultSet from JDBC driver and received the following error `java.sql.SQLException: Conversion from OBJECT not supported.`

The following is what the mapping of my index looks like:

    "mappings": {
      "properties": {
        "details": {
          "properties": {
            "memory": {
              "type": "long"
            },
            "model": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            }
          }
        },
        "item": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "price": {
          "properties": {
            "currency": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "value": {
              "type": "long"
            }
          }
        }
      }
    }

Result set has the following columns:
1. Column 1 of type VARCHAR and name: item
2. Column 2 of type STRUCT and Name: details
3. Column 3 of type STRUCT and Name: price

**How can one reproduce the bug?**
1. Create an index with a struct or nested struct
2. Execute a query using the JDBC driver
3. Attempt to extract data from ResultSet by calling `resultSet.getObject(<fieldName>)` or `resultSet.getObject(<fieldName>, <xxx.class>)`
4. See error mentioned above

**What is the expected behavior?**
It is expected to be able to retrieve the struct data type from the current row of the ResultSet, the same way other data types can be accessed.

**What is your host/environment?**
 - OS: Lambda Function with Java 11 (Corretto) Runtime
 - Version: JDBC Driver version 1.2.0.0

### PR DESCRIPTION
**Pull request by akuzin1**
### Description
Adds Struct support to driver.
 
### Issues Resolved
#65
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/jdbc/types/TypeConverters.java [hunk 1]
```diff
@@ -56,12 +57,37 @@ public class TypeConverters {
...
     public static TypeConverter getInstance(JDBCType jdbcType) {
         return tcMap.get(jdbcType);
     }
 
+    public static class StructTypeConverter extends BaseTypeConverter {
+
+        private static final Set<Class> supportedJavaClasses = Collections.unmodifiableSet(
+                new HashSet<>(Arrays.asList(
+                        Struct.class
+                )));
```
Comment (Yury-Fridlyand) [inline src/main/java/org/opensearch/jdbc/types/TypeConverters.java:74]: Maybe `... = Set.of(Struct.class);`? If it is available in this version of java.

**Review by Yury-Fridlyand — COMMENTED** [review]
Nice work!
You need to add unit tests, integration tests and fix DCO.

File: src/main/java/org/opensearch/jdbc/types/TypeConverters.java (hunk 1 above)
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/types/TypeConverters.java:74]: Yeah so for where we are using the JDBC driver we need java 8 support, so wouldn't be able to use the java 11 collections apis.

**Review by Yury-Fridlyand — APPROVED** [review]
Can you build the driver and run an OpenSearch cluster with corresponding data and verify that everything works fine now?
Few screenshots from a DB management tool showing **before** and **after** your fix would a great proof.
I can provide you a sample data for tests if you need.

**Comment by akuzin1** [comment]
Yeah I can definitely do that and any sample data would be great to have. Thank you.

**Comment by akuzin1** [comment]
Here are a couple screenshots documenting the behavior before and after the above change when using the sql jdbc driver to query OpenSearch index with structs.
**Before:**
Screenshot 2023-05-05 at 2 04 57 PM
**After:**
Screenshot 2023-05-05 at 1 20 01 PM

**Before:**
Screenshot 2023-05-05 at 2 02 26 PM
**After:**
Screenshot 2023-05-05 at 1 35 16 PM

File: src/main/java/org/opensearch/jdbc/ConnectionImpl.java [hunk 2]
```diff
@@ -41,6 +41,7 @@
 import java.sql.Savepoint;
 import java.sql.Statement;
 import java.sql.Struct;
+import java.util.Arrays;
```
Comment (acarbonetto) [inline src/main/java/org/opensearch/jdbc/ConnectionImpl.java:44]: is this needed?

File: src/main/java/org/opensearch/jdbc/StructImpl.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+package org.opensearch.jdbc;
+
+import java.sql.SQLException;
+import java.sql.Struct;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.List;
+
+public class StructImpl implements Struct{
```
Comment (acarbonetto) [inline src/main/java/org/opensearch/jdbc/StructImpl.java:15]:
```suggestion
public class StructImpl implements Struct {
```

File: src/main/java/org/opensearch/jdbc/StructImpl.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        Struct other = (Struct) obj;
+        try {
+            if (!typeName.equals(other.getSQLTypeName())) {
+                return false;
+            }
+            List otherAttributes = Arrays.asList(other.getAttributes());
+            return otherAttributes.containsAll(Arrays.asList(attributes));
+        }
+        catch (Exception e) {
```
Comment (acarbonetto) [inline src/main/java/org/opensearch/jdbc/StructImpl.java:52]: What are you trying to catch - we should avoid ignoring all `Exception`s

File: src/main/java/org/opensearch/jdbc/types/StructType.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    public String getTypeName() {
+        return "Struct";
+    }
+
+    @Override
+    public Struct fromValue(Object value, Map<String, Object> conversionParams) {
+        if (value == null) {
+          return null;
+        }
+        Map<String, Object> structKeyValues = (Map<String, Object>) value;
```
Comment (acarbonetto) [inline src/main/java/org/opensearch/jdbc/types/StructType.java:37]: is it possible this cast will fail?

**Review by acarbonetto — APPROVED** [review]

File: src/main/java/org/opensearch/jdbc/types/StructType.java (hunk 5 above)
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/types/StructType.java:37]: My experience showed that for structs it should not.

File: src/main/java/org/opensearch/jdbc/StructImpl.java (hunk 3 above)
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/StructImpl.java:15]: +1

File: src/main/java/org/opensearch/jdbc/ConnectionImpl.java (hunk 2 above)
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/ConnectionImpl.java:44]: Removed.

File: src/main/java/org/opensearch/jdbc/StructImpl.java (hunk 4 above)
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/StructImpl.java]: Supposed to be catching a SQLException, which I updated.

File: src/main/java/org/opensearch/jdbc/types/StructType.java (hunk 5 above)
Comment (acarbonetto) [reply src/main/java/org/opensearch/jdbc/types/StructType.java:37]: I'd consider adding a `instanceof` check in the `if` above.

File: src/main/java/org/opensearch/jdbc/types/StructType.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    }
+
+    @Override
+    public String getTypeName() {
+        return "Struct";
+    }
+
+    @Override
+    public Struct fromValue(Object value, Map<String, Object> conversionParams) {
```
Comment (dai-chen) [inline src/main/java/org/opensearch/jdbc/types/StructType.java:33]: Do you know what is the use of second param `conversionParams`?

File: src/main/java/org/opensearch/jdbc/types/BaseTypeConverter.java [hunk 7]
```diff
@@ -35,6 +36,9 @@ public abstract class BaseTypeConverter implements TypeConverter {
         typeHandlerMap.put(Date.class, DateType.INSTANCE);
         typeHandlerMap.put(Time.class, TimeType.INSTANCE);
 
+        // Adding Struct Support
```
Comment (dai-chen) [inline src/main/java/org/opensearch/jdbc/types/BaseTypeConverter.java:39]: np: the comment seems not very necessary. suggest to remove it. Same for the one in OpenSearchType.

Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/types/BaseTypeConverter.java]: Removed.

File: src/main/java/org/opensearch/jdbc/types/StructType.java (hunk 6 above)
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/types/StructType.java:33]: `conversionParams` is an optional parameter that gives the ability to pass in some additional metadata that can be used for creating value, currently used for most date/time types. However, it's not applicable for Struct type and many of the other classes that extend TypeHelper don't use it as well.

File: src/main/java/org/opensearch/jdbc/StructImpl.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+package org.opensearch.jdbc;
+
+import java.sql.SQLException;
+import java.sql.Struct;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.List;
+
+public class StructImpl implements Struct {
```
[the remark this answers is not in the document]
Comment (akuzin1) [reply src/main/java/org/opensearch/jdbc/StructImpl.java:22]: Added.

**Review by penghuo — APPROVED** [review]
Thanks for the change!

**Review by dai-chen — APPROVED** [review]
Thanks for the changes!

**Pull request merged**